### ISSUE TITLE
Apply patch to Flux kustomization using YAML manifest instead of inline patch

### ISSUE THREAD

**Issue by jeevankuduvaravindran** [fluxcd/flux2#4361]
### Describe the bug

I am looking to run a patch for helm release after a predeploy kubenertes job is completed. The current repository structure is similar to what has been defined in the flux docs [here](https://fluxcd.io/flux/use-cases/running-jobs/#repository-structure). Right now we don't have the ability to patch a Flux Kustomization with a yaml file. All we have is an inline patch which isn't clean enough for our standards. What if we need to run the patch only after running a pre job by depending on another kustomization and apply the patch only after the pre job is completed? It will be very useful to have a yaml reference instead of inline patch. This is because since the patch.yaml is looking for a helmrelease that doesn't exist until predeploy is completed and is available only when deploy is started.  something like below will be very helpful.

```
apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  name: demo-deploy
  namespace: config-cluster-flux
spec:
  dependsOn:
    - name: demo-pre-deploy
      namespace: config-cluster-flux
  sourceRef:
    kind: GitRepository
    name: services-repository
    namespace: config-cluster-flux
  path: ./services/demo/deploy
  interval: 5m
  timeout: 5m
  prune: true
  wait: true
  patches:
    - path: patch.yaml
```

Patch.yaml has the patch for a helmrelease that is available as part of deploy kustomization. 

Is there any other way we could achieve this? or am I missing something? Any help on this would be much appreciated. 

Thanks!

### Steps to reproduce

1. Follow the same procedure as setting up a db migration job with pre-deploy and deploy stages as provided in the documentation below
https://fluxcd.io/flux/use-cases/running-jobs/#repository-structure
2. Apply an environment and cluster level patch for the helmrelease in deploy stage

### Expected behavior

Patch should be applied from a yaml file instead of inline patch in deploy kustomize file (Flux kustomization)

### OS / Distro

N/A

### Flux version

v2.1.0

### Flux check

► checking prerequisites
✗ flux 2.1.0 <2.1.2 (new version is available, please upgrade)
✔ Kubernetes 1.26.6 >=1.25.0-0
► checking controllers
✔ fluxconfig-agent: deployment ready
► mcr.microsoft.com/azurek8sflux/fluxconfig-agent:1.8.0
► mcr.microsoft.com/azurek8sflux/fluent-bit-mdm:1.8.0
✔ fluxconfig-controller: deployment ready
► mcr.microsoft.com/azurek8sflux/fluxconfig-controller:1.8.0
► mcr.microsoft.com/azurek8sflux/fluent-bit-mdm:1.8.0
✔ helm-controller: deployment ready
► mcr.microsoft.com/oss/fluxcd/helm-controller:v0.36.1
✔ kustomize-controller: deployment ready
► mcr.microsoft.com/oss/fluxcd/kustomize-controller:v1.1.0
✔ notification-controller: deployment ready
► mcr.microsoft.com/oss/fluxcd/notification-controller:v1.1.0
✔ source-controller: deployment ready
► mcr.microsoft.com/oss/fluxcd/source-controller:v1.1.1
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta2
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ fluxconfigs.clusterconfig.azure.com/v1alpha1
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta2
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta2
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta2
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by stefanprodan** [comment]
Only in-line patches are supported in Flux Kustomizations.

**Comment by jeevankuduvaravindran** [comment]
@stefanprodan Thanks for your quick response. Do you suggest any other way we could do a patch on helm release from the deploy stage? We tried a postBuild substitute for patch from kustomize Kustomization like below and it too didn't work. 

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - helm-repository.yaml
  - helm-release.yaml
  - image-repository.yaml  
  - image-policy.yaml

patches:
  - path: ../${ENVIRONMENT}-${CLUSTER}.yaml
```

**Comment by kingdonb** [comment]
If you're trying to patch the resources inside of a `HelmRelease` from the Kustomization that applies the HR, that will not work, the Kustomization cannot "look through" the HelmRelese, they are reconciled in different controllers.

https://fluxcd.io/flux/components/helm/helmreleases/#post-renderers

However the Helm Controller also has a postrender kustomization feature that you can use to edit rendered resources once `helm template` stage is done, this is a built-in feature of Helm that Flux supports. You still cannot use a file from the filesystem, the `HelmRelease` or `Kustomization` does not have access to the filesystem surrounding it at build/patch/apply time, it's not in any directory, it is in the cluster and etcd, so any patches must come from the spec. Hope this helps!

**Comment by mavogel** [comment]
Note: duplicate of https://github.com/fluxcd/flux2/issues/4030. Had the same request as inline patches gone too long
